discovery/README.md

```diff
@@ -146,85 +146,132 @@ both cases.
-       },
-    },
-    Labels: map[LabelName][LabelValue] {
-      "job": "postgres",
-    },
-    "Source": "file2", 
-  },
+	{
+		Targets: []model.LabelSet{
+			{
+				"__instance__": "10.11.150.1:7870",
+				"hostname":     "demo-target-1",
+				"test":         "simple-test",
+			},
+			{
+				"__instance__": "10.11.150.4:7870",
+				"hostname":     "demo-target-2",
+				"test":         "simple-test",
+			},
+		},
+		Labels: model.LabelSet{
+			"job": "mysql",
+		},
+		"Source": "file1",
+	},
+	{
+		Targets: []model.LabelSet{
+			{
+				"__instance__": "10.11.122.11:6001",
+				"hostname":     "demo-postgres-1",
+				"test":         "simple-test",
+			},
+			{
+				"__instance__": "10.11.122.15:6001",
+				"hostname":     "demo-postgres-2",
+				"test":         "simple-test",
+			},
+		},
+		Labels: model.LabelSet{
+			"job": "postgres",
+		},
+		"Source": "file2",
+	},
 }
 ```
 
 Here there are two target groups one group with source `file1` and another with `file2`. The grouping is implementation specific and could even be one target per group. But, one has to make sure every target group sent by an SD instance should have a `Source` which is unique across all the target groups of that SD instance. 
 
 In this case, both the target groups are sent down the channel the first time `Run()` is called. Now, for an update, we need to send the whole _changed_ target group down the channel. i.e, if the target with `hostname: demo-postgres-2` goes away, we send:
-```
+```go
 &targetgroup.Group{
-  Targets: []model.LabelSet{
-     {
-        "__instance__": "10.11.122.11:6001",
-        "hostname": "demo-postgres-1",
-        "test": "simple-test",
-     },
-  },
-  Labels: map[LabelName][LabelValue] {
-    "job": "postgres",
-  },
-  "Source": "file2", 
+	Targets: []model.LabelSet{
+		{
+			"__instance__": "10.11.122.11:6001",
+			"hostname":     "demo-postgres-1",
+			"test":         "simple-test",
+		},
+	},
+	Labels: model.LabelSet{
+		"job": "postgres",
+	},
+	"Source": "file2",
 }
 ```
 down the channel.
 
 If all the targets in a group go away, we need to send the target groups with empty `Targets` down the channel. i.e, if all targets with `job: postgres` go away, we send:
-```
+```go
 &targetgroup.Group{
-  Targets: nil,
-  "Source": "file2", 
+	Targets:  nil,
+	"Source": "file2",
 }
 ```
 down the channel.
 
+### The Config interface
+
+Now that your service discovery mechanism is ready to discover targets, you must help
+Prometheus discover it. This is done my implementing the `discovery.Config` interface
+and registering it with `config.RegisterServiceDiscovery` in an init function of your
+package.
+
+```go
+type Config interface {
+	// Name returns the name of the discovery mechanism.
+	Name() string
+
+	// NewDiscoverer returns a Discoverer for the Config
+	// with the given DiscovererOptions.
+	NewDiscoverer(DiscovererOptions) (Discoverer, error)
+
+	// SetOptions applies the ConfigOptions to the Config.
+	SetOptions(ConfigOptions)
```

Comment:
Neither SetOptions not Validate should be here. We don't need that option, and any validation should be done as part of unmarshalling - not as a separate step.

Reply:
SetOptions is there to preserve the existing behavior.

I added Validate during the review to support `promtool`, whose existing behavior was only validating kubernetes configs.

Reply:
Why can't this go in NewDiscoverer?

Reply:
By the time NewDiscoverer is called, the context of which directory the config file was loaded from is long gone.

FWIW, the resolving of paths relative to the config file's path isn't just for service discovery configs. It also happens in:
- rule files
- scrape configs (http client config)
- alertmanager configs (http client config)
- remote read configs (http client config)
- remote write configs (http client config)

If you're still set on changing this behavior, it can be a separate issue. Since there are no API guarantees, leaving SetOptions shouldn't be a problem for now.

Reply:
> By the time NewDiscoverer is called, the context of which directory the config file was loaded from is long gone.

Why not plumb it through?

> If you're still set on changing this behavior, it can be a separate issue. Since there are no API guarantees, leaving SetOptions shouldn't be a problem for now.

Not having API guarantees means changes can break APIs, it doesn't mean that code health can be ignored.

Reply:
If we delay it until `NewDiscoverer`, then the configs can't be validated before use.

I understand the desire to keep the original config pristine for rendering. However, plumbing the original config's directory through to every use of rule files, scrape configs, alertmanager configs, remote read configs, remote write configs, and service discovery configs seems much less maintainable/healthy than the status quo of the last 5 years. The current unmarshaling contains many examples of overriding values with defaults, copying global config values into scrape configs, etc. The config served by the web API just isn't the same as the one on disk. Arguably more importantly, it's how Prometheus is interpreting the config on disk, which includes how it resolves file paths.

Reply:
Here's a sketch of a DirectorySetter interface implemented recursively with reflection, deep cloning implemented recursively with reflection, all to save the original config for the web api: https://github.com/abursavich/prometheus/compare/config...abursavich:setdir

Reply:
> If we delay it until NewDiscoverer, then the configs can't be validated before use.

We could catch it during reload, but I have a strong suspicion that the reload code doesn't generally handle that correctly currently.

> seems much less maintainable/healthy than the status quo of the last 5 years.

I'm not sure there. We've had the same bugs again and again where a new field wasn't added to the various places it needs to be added for all this to be handled correctly. That sort of accidental oversight needs to be dealt with, so something that you can't forget to implement would be an improvement. So if a SetDirectory approach is used, it needs to be either a panic or compile failure not to implement and call it.

> The current unmarshaling contains many examples of overriding values with defaults, copying global config values into scrape configs, etc. The config served by the web API just isn't the same as the one on disk.

That is a good point.

> deep cloning implemented recursively with reflection

That reflection doesn't feel right at all, we should be using explicit interfaces everywhere we can (which is basically everywhere bar where the SDs dynamically hook in).

Reply:
re: `SetDirectory`

My original sketch for `Clone` and `SetDirectory` both used reflection because I didn't trust that people would keep the `Clone`/`SetDirectory` methods up to date with changes to their types (e.g. new fields).

The reflection version of `SetDirectory` would recurse through the values calling `SetDirectory` on any type with the method. If your config struct had an `HTTPClientConfig` field, it would find the `HTTPClientConfig` and call its `SetDirectory` method without your type needing to implement a proxy. However, if you did implement `SetDirectory` on your type it would trust that you knew what you were doing and not recurse deeper.

I ended up making `SetDirectory` explicit and keeping reflection for `Clone`, since I generally think explicit is better than reflection (magic) and the number of things people have to track and get right for `SetDirectory` is much smaller than for `Clone`.

It's worth noting that `SetDirectory` and `Validate` will generally have the same shape and should probably use the same calling convention if both are implemented. Things that are known at `Unmarshal` won't need `SetDirectory` or `Validate`.

I could switch back to reflection for `SetDirectory` or keep it as is and use reflection in tests to make sure `SetDirectory` has been called on all of the "leaves"?

Side Note: The current behavior doesn't update QueryLogFile in GlobalConfig.

---

re: `Clone`

I only needed to add `Clone` to keep a separate copy of the config for the web API. If we drop that extra requirement for now, since there are many other areas where it doesn't match what's on disk, we can get rid of `Clone`.

Side Note: It would be simple to use the raw file data for the web API, but that would expose secrets which makes it a no-go.

---

I'm having a hard time keeping track of what the criteria is for getting this PR though. I would be happy to contribute to some of these other issues, but I would like to avoid this becoming a catch all for everything we want to change about configs. I think it makes sense to come to an agreement on what to do about `SetDirectory` and to spin off config validation and original config preservation to separate issues.

Reply:
> I would like to avoid this becoming a catch all for everything we want to change about configs. I think it makes sense to come to an agreement on what to do about SetDirectory and to spin off config validation and original config preservation to separate issues.

This seems reasonable to me.

Reply:
Here's `SetDirectory` where it's explicitly called down the value chain: https://github.com/abursavich/prometheus/compare/config...abursavich:setdir-v2

There's already some decent testing with the existing "good" test config, but it only cover fields that are used (e.g. values that have an `HTTPClientConfig` must specify paths in the testdata for it to be useful here). I added reflection-based testing that will take the "good" config  and set any field that looks like a file path (e.g. its name has the suffix `File` or `Files`).

Reply:
> I ended up making SetDirectory explicit and keeping reflection for Clone, since I generally think explicit is better than reflection (magic) 

I agree, let's avoid reflection where we can.

> I could switch back to reflection for SetDirectory or keep it as is

As-is I'd say.

> use reflection in tests to make sure SetDirectory has been called on all of the "leaves"?

I'd prefer to find a way that doesn't need it. For example can the HTTPClientConfig fail in some way if SetDirectory is not called either at Validate or instantiation? A explicit failure will be caught and easily fixed (even if it manages to make it out into a release), whereas silently not setting it will not be easy to spot or debug.

> The current behavior doesn't update QueryLogFile in GlobalConfig.

That is a bug we should fix at some point then.

> I only needed to add Clone to keep a separate copy of the config for the web API.

We don't need that it looks like.

> I'm having a hard time keeping track of what the criteria is for getting this PR though. 

I can appreciate that. At this point we're still figuring out how to fully design this, so some back and forth discussion is inevitable as this is a non-trivial problem. This also involves paying off some technical debt that has built up over the years, as we can no longer cheat with layering violations for validation.

> to spin off config validation and original config preservation to separate issues.

We can't break existing functionality, and both config validation and showing the config in the UI are existing features.

>  I added reflection-based testing that will take the "good" config and set any field that looks like a file path

Keep in mind that Prometheus is not the only thing using HTTPClientConfig, and reflection-based testing is enough of a smell already without having to replicate it to other repositories.

Reply:
> > spin off config validation and original config preservation to separate issues
> 
> We can't break existing functionality, and both config validation and showing the config in the UI are existing features.

The current version of this PR preserves the existing functionality (or lack thereof) regarding validation (same `Unmarshal` checks) and preservation (file paths were and are modified), which is why I suggest spinning them off to separate issues.

> Keep in mind that Prometheus is not the only thing using HTTPClientConfig, and reflection-based testing is enough of a smell already without having to replicate it to other repositories.

> For example can the HTTPClientConfig fail in some way if SetDirectory is not called either at Validate or instantiation? A explicit failure will be caught and easily fixed (even if it manages to make it out into a release), whereas silently not setting it will not be easy to spot or debug.

This is tricky... There are `Load` functions and `LoadFile` functions and only `LoadFile` sets the directory. Depending on the different use cases, calling it might not be required? It might make sense to construct and use an HTTPClientConfig that isn't read directly from a file?

I just looked at the alertmanager config code and it has plenty of `HTTPClientConfig`s, but [it only updates `Templates`](https://github.com/prometheus/alertmanager/blob/8e861c646bf67599a1704fc843c6a94d519ce312/config/config.go#L189-L217). This is probably a bug, but what to do about it now?

A nice thing about the reflection testing is it confirms that calling SetDirectory on a root value is propagated to its inner and leaf values, but it doesn't force you to into calling SetDirectory on anything. The opt in nature of this testing gives other components (e.g. alertmanager) time to sort out their behavior without failing at runtime.

I factored the testing code out into `common/config/configtest` so it can be easily adopted by other repos.

Reply:
> The current version of this PR preserves the existing functionality (or lack thereof) 

Not going backwards is the main thing. We can no longer reach into the internals of HTTPClientConfig from Prometheus though.

> It might make sense to construct and use an HTTPClientConfig that isn't read directly from a file?

It doesn't, they all come from either user-provided config files or unittests. Even then you could pass in pwd or whatnot if they came from elsewhere. We could for example add the base directory as a argument to Load - though unittests tend to instantiate a HTTPClientConfig directly.

> This is probably a bug, but what to do about it now?

Yip, that's a bug. Not a problem for this PR, but will need resolving when this change is propagated out. Current problems like this need to be fixed systemically, even if not all in this PR.

> it doesn't force you to into calling SetDirectory on anything. The opt in nature of this testing gives other components (e.g. alertmanager) time to sort out their behavior without failing at runtime.

We control when the alertmanager picks up a new version of common, and I'd expect that all to be done in one PR. I'd really prefer not to have that sort of reflection lying around the codebase.

Reply:
Okay... So I think we're just down to how to test/enforce that the code properly propagates calls to `SetDirectory`.

I prefer the reflection-based tests because they cover all `DirectorySetters` (e.g. `marathon.SDConfig` which includes an `AuthTokenFile`) and are now easy to add to any root config you want to test. Since tests instantiate `HTTPClientConfig` and `TLSConfig` directly for comparison, it would be burdensome to make them flip some unexported `hasSetDir` bool when they already use the paths themselves to verify that `testdata/` has been properly prepended. Also, every instance where `HTTPClientConfig` or `TLSConfig` is empty/unused in tests would require the same bit flip. Practically, that seems like a lot of tedium and headache for little gain.

Reply:
Once I created `common/config/configtest`, this is what was required to add the test to `prometheus/config`, and it would be similar for others like `alertmanager/config`:

```go
func TestSetDirectory(t *testing.T) {
	files, err := filepath.Glob("testdata/*.good.yml")
	testutil.Ok(t, err)
	testutil.Assert(t, len(files) > 0, "no good configs found")

	loadFile := func(file string) (config_util.DirectorySetter, error) {
		return LoadFile(file)
	}
	for _, file := range files {
		configtest.TestSetDirectory(t, file, loadFile, assertConfigsEqual)
	}
}
```

Reply:
> it would be burdensome to make them flip some unexported hasSetDir bool when they already use the paths themselves to verify that testdata/ has been properly prepended

It'd be far, far from the first time we needed to bulk change tests for something like that. We could also work off the paths being relative rather than absolute rather than an explicit flag.

> this is what was required to add the test to prometheus/config,

It's still a whole pile of reflection backing it, and I'd rather not have that reflection.

Reply:
I'm still seeing that reflection.

Reply:
> I'm still seeing that reflection.

Heh, yep :smile:

I also prefer not to use reflection when it isn't necessary, but its use in these tests covers more than would be by other means. It makes sure `SetDirectory` is propagated to any inner/leaf values that implement it and its heuristics find values like `QueryLogFile` in `GlobalConfig` that are files but aren't updated. The former of which may be left out of hand-rolled tests/validation and the later of which would never be covered by hand-rolled tests/validation.